### PR TITLE
[Alpha pipeline] Add captioning component

### LINKED ISSUE #98: Captioning component

### PR DESCRIPTION
**Pull request by NielsRogge**
This PR adds a generic image captioning component.

Fixes #98

### CODE REVIEW

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (RobbeSneyders) [inline examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:105]:
```suggestion
    component = CaptionImagesComponent.from_file()
```

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (RobbeSneyders) [inline examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:1]: Would love to see if we can translate this to the dataframe interface in the future, but for now this is fine for me.

**Review by RobbeSneyders — COMMENTED** [review]
Thanks @NielsRogge!

LGTM apart from some small comments. Would like to get @PhilippeMoussalli's review on the Dask implementation as well though.

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (PhilippeMoussalli) [inline examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:16]: Can you group the 3rd party imports together?

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (PhilippeMoussalli) [inline examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:46]: I think the `max_length` should also be a component parameter

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (PhilippeMoussalli) [inline examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:93]: does this also need to be delayed? Have you tried passing it without?

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (PhilippeMoussalli) [inline examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:101]: Does this dataframe contain the ids?

**Review by PhilippeMoussalli — COMMENTED** [review]
Thanks Niels! This will be a nice example for future components that involve inferencing with a model.
Left a few comments

**Comment by RobbeSneyders** [comment]
Do we actually need this component for the ControlNet pipeline? We already have captions from LAION, no?

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (bertChristiaens1) [inline examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:86]: Have we tested this with larger amounts of data? It seems that e.g. naively loading 1M+ images into memory could use a lot of RAM

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (bertChristiaens1) [inline examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:88]: Also here, moving all the images to the device could be problematic (certainly if device='cuda'). This would give CUDA Out of memory

File: examples/pipelines/controlnet-interior-design/run_locally.sh
Comment (RobbeSneyders) [inline examples/pipelines/controlnet-interior-design/run_locally.sh:1]: Nice, but I assume this isn't meant to be included with your hard-coded arguments?

Would be great if the pipeline could do this, then we don't even need a kfp cluster locally. Not sure how this would work with hardware requirements though.

File: components/prompt_based_laion_retrieval/src/main.py
Comment (RobbeSneyders) [inline components/prompt_based_laion_retrieval/src/main.py:35]: Why is this try except needed? Would prefer to at least not have a bare `except` (and I'm surprised ruff doesn't complain about this)

File: components/prompt_based_laion_retrieval/src/main.py
Comment (RobbeSneyders) [inline components/prompt_based_laion_retrieval/src/main.py:94]: Why are you resetting the index here?

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (RobbeSneyders) [inline examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:85]: These `len()` calls trigger a compute, which will have a high cost when working with larger than memory data. We probably want to include logging the input and output size lazily in the component class.

**Review by RobbeSneyders — COMMENTED** [review]
Thanks @NielsRogge!

File: components/prompt_based_laion_retrieval/src/main.py
Comment (NielsRogge) [reply components/prompt_based_laion_retrieval/src/main.py:74]: Yes, although logging complains when providing non-string objects like `dataframe.head(2)` to see the first few rows

File: components/prompt_based_laion_retrieval/src/main.py
Comment (NielsRogge) [reply components/prompt_based_laion_retrieval/src/main.py:94]: Because @GeorgesLorre asked me to do so :D

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (NielsRogge) [reply examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:85]: Yes I know, its mainly for debugging

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (PhilippeMoussalli) [inline examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:86]: would remove those print statements before merging

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (PhilippeMoussalli) [inline examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:120]: same here

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (GeorgesLorre) [inline examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:118]: the number of rows does not change in this component right ? We only add columns ?

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (GeorgesLorre) [inline examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:86]: Could you remove the print statements or convert them to logs?

File: examples/pipelines/controlnet-interior-design/pipeline.py
Comment (PhilippeMoussalli) [inline examples/pipelines/controlnet-interior-design/pipeline.py:49]: would make the batch larger before merging

File: examples/pipelines/controlnet-interior-design/run_locally.sh
Comment (GeorgesLorre) [inline examples/pipelines/controlnet-interior-design/run_locally.sh:1]: I would not merge this into main but it is a nice script we can start improving to speed up local development

Comment (PhilippeMoussalli) [inline examples/pipelines/controlnet-interior-design/run_locally.sh:1]: I'm a bit conflicted about having this merged on the main branch.
Do we want to set an example/way of working on running things locally like this @GeorgesLorre ? 
Maybe leave out your name from it and just keep some defaults

**Review by PhilippeMoussalli — APPROVED** [review]
Approved! Thanks for implementing this :) 
Left a few minor comments

File: examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py
Comment (GeorgesLorre) [reply examples/pipelines/controlnet-interior-design/components/caption_images/src/main.py:115]: Then we should not have to drop the index in this case since it does not change

**Pull request merged**
